File: convert-docx.js
const fs = require('fs');
const path = require('path');

// Проверяем и устанавливаем библиотеку mammoth
let mammoth;
try {
    mammoth = require('mammoth');
    console.log('mammoth library found');
} catch (error) {
    console.log('Installing mammoth library...');
    const { execSync } = require('child_process');
    
    try {
        execSync('npm install mammoth', { stdio: 'inherit' });
        mammoth = require('mammoth');
        console.log('mammoth library installed successfully');
    } catch (installError) {
        console.error('ERROR: Cannot install mammoth. Please install manually: npm install mammoth');
        process.exit(1);
    }
}

async function extractTextFromDocx(docxPath) {
    try {
        const result = await mammoth.extractRawText({ path: docxPath });
        return result.value.trim();
    } catch (error) {
        return `ERROR reading document: ${error.message}`;
    }
}

async function main() {
    const ecoveryPath = 'эковери';
    const outputPath = 'товары_данные';
    
    // Создаем папку для результатов
    if (!fs.existsSync(outputPath)) {
        fs.mkdirSync(outputPath, { recursive: true });
    }
    
    if (!fs.existsSync(ecoveryPath)) {
        console.error(`ERROR: Directory '${ecoveryPath}' not found`);
        return;
    }
    
    // Получаем все папки товаров
    const productFolders = fs.readdirSync(ecoveryPath, { withFileTypes: true })
        .filter(dirent => dirent.isDirectory())
        .map(dirent => dirent.name);
    
    console.log(`Found ${productFolders.length} product folders`);
    
    let processedCount = 0;
    
    for (const productName of productFolders) {
        console.log(`Processing: ${productName}`);
        
        const productPath = path.join(ecoveryPath, productName);
        
        // Ищем .docx файлы в папке товара
        const files = fs.readdirSync(productPath);
        const docxFiles = files.filter(file => file.endsWith('.docx'));
        
        if (docxFiles.length === 0) {
            console.log(`  WARNING: No .docx files found in ${productName}`);
            
            // Создаем пустой файл для товара без документа
            const outputFile = path.join(outputPath, `${productName}.txt`);
            const emptyContent = [
                `Product: ${productName}`,
                `Price: NO DATA`,
                `Description: NO DATA`,
                `Status: Missing description file`
            ].join('\n');
            
            fs.writeFileSync(outputFile, emptyContent, 'utf8');
            processedCount++;
            continue;
        }
        
        // Обрабатываем первый .docx файл
        const docxFile = docxFiles[0];
        console.log(`  Converting: ${docxFile}`);
        
        try {
            const docxPath = path.join(productPath, docxFile);
            
            // Извлекаем текст из документа
            const textContent = await extractTextFromDocx(docxPath);
            
            // Создаем структурированный контент
            const outputFile = path.join(outputPath, `${productName}.txt`);
            const structuredContent = [
                `Product: ${productName}`,
                `Source file: ${docxFile}`,
                `Content:`,
                textContent
            ].join('\n');
            
            fs.writeFileSync(outputFile, structuredContent, 'utf8');
            
            console.log(`  ✓ Saved to: ${outputFile}`);
            processedCount++;
            
        } catch (error) {
            console.error(`  ERROR processing ${docxFile}: ${error.message}`);
            
            // Создаем файл с ошибкой
            const outputFile = path.join(outputPath, `${productName}.txt`);
            const errorContent = [
                `Product: ${productName}`,
                `Source file: ${docxFile}`,
                `Status: CONVERSION ERROR`,
                `Error: ${error.message}`
            ].join('\n');
            
            fs.writeFileSync(outputFile, errorContent, 'utf8');
            processedCount++;
        }
    }
    
    console.log('\n=== DONE ===');
    console.log(`Total products processed: ${processedCount}`);
    console.log(`Files saved to: ${outputPath}`);
    
    // Показываем список созданных файлов
    const txtFiles = fs.readdirSync(outputPath)
        .filter(file => file.endsWith('.txt'))
        .sort();
    
    console.log(`\nCreated ${txtFiles.length} files:`);
    txtFiles.forEach(file => {
        console.log(`  - ${file}`);
    });
}

// Запускаем скрипт
main().catch(error => {
    console.error('Script error:', error);
    process.exit(1);
});

File: convert_docx.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import os
import sys
from pathlib import Path

# Проверяем установлена ли python-docx
try:
    from docx import Document
    print("python-docx library found")
except ImportError:
    print("Installing python-docx library...")
    os.system("pip install python-docx")
    try:
        from docx import Document
        print("python-docx library installed successfully")
    except ImportError:
        print("ERROR: Cannot install python-docx. Please install manually: pip install python-docx")
        sys.exit(1)

def extract_text_from_docx(docx_path):
    """Extract text from Word document"""
    try:
        doc = Document(docx_path)
        text_content = []
        
        for paragraph in doc.paragraphs:
            if paragraph.text.strip():
                text_content.append(paragraph.text.strip())
        
        return '\n'.join(text_content)
    except Exception as e:
        return f"ERROR reading document: {str(e)}"

def main():
    ecovery_path = Path("эковери")
    output_path = Path("товары_данные")
    
    # Create output directory
    output_path.mkdir(exist_ok=True)
    
    if not ecovery_path.exists():
        print(f"ERROR: Directory '{ecovery_path}' not found")
        return
    
    # Get all product folders
    product_folders = [f for f in ecovery_path.iterdir() if f.is_dir()]
    print(f"Found {len(product_folders)} product folders")
    
    processed_count = 0
    
    for folder in product_folders:
        product_name = folder.name
        print(f"Processing: {product_name}")
        
        # Find .docx files in the folder
        docx_files = list(folder.glob("*.docx"))
        
        if not docx_files:
            print(f"  WARNING: No .docx files found in {product_name}")
            
            # Create empty file for products without documents
            output_file = output_path / f"{product_name}.txt"
            with open(output_file, 'w', encoding='utf-8') as f:
                f.write(f"Product: {product_name}\n")
                f.write("Price: NO DATA\n")
                f.write("Description: NO DATA\n")
                f.write("Status: Missing description file\n")
            
            processed_count += 1
            continue
        
        # Process the first .docx file
        docx_file = docx_files[0]
        print(f"  Converting: {docx_file.name}")
        
        try:
            # Extract text from document
            text_content = extract_text_from_docx(docx_file)
            
            # Create structured content
            output_file = output_path / f"{product_name}.txt"
            with open(output_file, 'w', encoding='utf-8') as f:
                f.write(f"Product: {product_name}\n")
                f.write(f"Source file: {docx_file.name}\n")
                f.write("Content:\n")
                f.write(f"{text_content}\n")
            
            print(f"  ✓ Saved to: {output_file}")
            processed_count += 1
            
        except Exception as e:
            print(f"  ERROR processing {docx_file.name}: {str(e)}")
            
            # Create error file
            output_file = output_path / f"{product_name}.txt"
            with open(output_file, 'w', encoding='utf-8') as f:
                f.write(f"Product: {product_name}\n")
                f.write(f"Source file: {docx_file.name}\n")
                f.write("Status: CONVERSION ERROR\n")
                f.write(f"Error: {str(e)}\n")
            
            processed_count += 1
    
    print("\n=== DONE ===")
    print(f"Total products processed: {processed_count}")
    print(f"Files saved to: {output_path}")
    
    # List created files
    txt_files = list(output_path.glob("*.txt"))
    print(f"\nCreated {len(txt_files)} files:")
    for txt_file in sorted(txt_files):
        print(f"  - {txt_file.name}")

if __name__ == "__main__":
    main() 